runtime/compiler/optimizer/InterpreterEmulator.hpp

```diff
@@ -0,0 +1,272 @@
+/*******************************************************************************
+ * Copyright (c) 2000, 2019 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+/*
+ * \class InterpreterEmulator
+ *
+ * \brief This class is a bytecode iterator in estimate code size (ECS) of inliner.
+ *
+ * \notes The iterator has statelss and with state modes.
+ *        With state mode is used to emulate the interpreter execution with
+ *        an operand stack maintained during bytecode iteration to keep useful
+ *        information like known object and constant integer so
+ *        that inliner can make better decision when creating callsites. For
+ *        example, inliner can avoid creating callsites on dead path
+ *        \ref maintainStackForIf or can refine callee method based on known
+ *        receiver info \ref refineResolvedCalleeForInvokestatic. Operands that
+ *        can't be reasoned about are represented by a dummy operand \ref _unknownOperand
+ *        which doesn't carry any extra information. Currently, with state mode only
+ *        supports methodhandle thunk archetypes.
+ */
+
+#ifndef INTERPRETER_EMULATOR_INCL
+#define INTERPRETER_EMULATOR_INCL
+
+#include "il/Block.hpp"
+#include "ilgen/ByteCodeIteratorWithState.hpp"
+#include "ilgen/J9ByteCodeIterator.hpp"
+#include "compile/Compilation.hpp"
+#include "optimizer/Inliner.hpp"
+#include "optimizer/J9Inliner.hpp"
+#include "optimizer/J9EstimateCodeSize.hpp"
+#include "env/TRMemory.hpp"
+
+class IconstOperand;
+class KnownObjOperand;
+class MutableCallsiteTargetOperand;
+
+class Operand
```

Comment:
This feels like it should be abstract, but nothing is forcing it to be such... At the very least some doxygen is required. Should there be a dump/print operation on these things to make logging easier?

Reply:
Added doxygen comments to explain `Operand` and `printToString` for easy logging. I didn't make it abstract because we also need to instantiate a general Operand object to represent the unknown object.